File: 2197-replace-non-coprime-numbers-in-array/stack/python/solution.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        for num in nums:
            stack.append(num)

            while len(stack) >= 2:
                num1, num2 = stack[-1], stack[-2]

                hcf = math.gcd(num1, num2)
                if hcf == 1: # co-prime, no merging needed
                    break

                lcm = num1 * num2 // hcf
                stack.pop()
                stack.pop()
                stack.append(lcm)
                
        return stack

File: 2197-replace-non-coprime-numbers-in-array/stack/java/solution.java
class Solution {
    public List<Integer> replaceNonCoprimes(int[] nums) {
        List<Integer> stack = new ArrayList<>();
        for (int num : nums) {
            stack.add(num);

            while (stack.size() >= 2) {
                int num1 = stack.get(stack.size() - 1);
                int num2 = stack.get(stack.size() - 2);
                
                int hcf = gcd(num1, num2);
                if (hcf == 1) { // co-prime, no merging needed
                    break;
                }

                long lcm = ((long) num1 * num2) / hcf;
                stack.remove(stack.size() - 1);
                stack.remove(stack.size() - 1);
                stack.add((int) lcm);
            }
        }
        return stack;
    }
    
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}
